plan/heuristic_planner.go

```diff
@@ -51,8 +53,11 @@ func (p *heuristicPlanner) matchRules(ctx context.Context, node Node) (Node, boo
 			newNode, changed, err := rule.Rewrite(ctx, node)
 			if err != nil {
 				return nil, false, err
+			} else if !changed {
+				continue
```

Comment:
This looks like an error, but I could be reading it wrong. By continuing here the `node = newNode` statement is not executed when it would have been in the previous code. Is there a reason that's different now?  If this is indeed an error then why didn't tests fail? Can you add a test case to cover that path if needed?

Reply:
Hm, my understanding was that if `changed` was false, the rewrite rule did not run and nothing should be done. It seems like all of the planner rules just happened to always return their own selves whenever they returned false rather than nil or anything else.

Reply:
I wasn't really sure how to change this so I just reverted the behavior to what it was previously.